### DISCUSSION TITLE
Mock and reset spy multiple times in one test

### DISCUSSION THREAD

**Discussion by zirkelc** [vitest-dev/vitest#6739] [General]
I'm using `vi.spyOn()` to spy or mock instance methods, for example `database.query() calls`. In one test file, I have several tests and some tests should be able to execute the spied function `database.query()` and I want to assert the invoked parameters, whereas in other test cases, I want to simply mock the returned value of `database.query()` without executing a real database query.

I noticed that if I use `vi.spyOn(Database.protoype, 'query').mockResolvedValue()` once, and the reset it via `vi.restoreAllMocks()` or calling `mockReset()` on the spy, it won't reset it to the actual implementation of calling the `database.query()` function.

This is clearly described in the docs, so my obviously my bad:
https://vitest.dev/api/vi.html#vi-spyon

But I was still wondering what is the actual reason you cannot mock and reset a spy multiple times in one test file?

**Comment by hi-ogawa** [comment]
Probably this is what you're looking for https://github.com/vitest-dev/vitest/pull/6426
We also think current behavior isn't intuitive, but it's a breaking change, so it's planned for next major release.

**Reply by zirkelc** [reply]
Awesome, thank you!
